app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -224,8 +232,13 @@ public void reportProgress(int progress) {
                         wasKeyLoggedIn = true;
                     }
 
-
                     this.publishProgress(PROGRESS_AUTHED, 0);
+                    if (isCancelled()) {
+                        // About to enter data commit phase; last chance to finish early if cancelled
+                        return PullTaskResult.UNKNOWN_FAILURE;
+                    }
+                    this.publishProgress(PROGRESS_DOWNLOADING_COMPLETE, 0);

```

Comment:
I think it's fine to not specifically articulate what action should happen, v. knowing that downloading is complete disables the button